presto-main/src/main/java/com/facebook/presto/operator/scalar/GroupingOperationFunction.java

```diff
@@ -0,0 +1,85 @@
+
+public final class GroupingOperationFunction
+{
+    private GroupingOperationFunction() {}
+
+    /**
+     * The grouping function is used in conjunction with GROUPING SETS, ROLLUP and CUBE to
+     * indicate which columns are present in that grouping.
+     *
+     * <p>The grouping function must be invoked with arguments that exactly match the columns
+     * referenced in the corresponding GROUPING SET, ROLLUP or CUBE clause at the associated
+     * query level. Those column arguments are not evaluated and instead the function is
+     * re-written with the arguments below.
+     *
+     * <p>To compute the resulting bit set for a particular row, bits are assigned to the
+     * argument columns with the rightmost column being the most significant bit. For a
+     * given grouping, a bit is set to 0 if the corresponding column is included in the
+     * grouping and 1 otherwise. For an example, see the SQL documentation for the
+     * function.
+     *
+     * @param groupId An ordinal indicating which grouping is currently being processed.
+     *        Each grouping is assigned a unique monotonically increasing integer.
+     * @param groupingOrdinals The column arguments with which the function was
+     *        invoked converted to ordinals with respect to the base table column
+     *        ordering.
+     * @param groupingSetOrdinals A collection of ordinal lists where the index of
+     *        the list is the groupId and the list itself contains the ordinals of the
+     *        columns present in the grouping. For example: [[0, 2], [2], [0, 1, 2]]
+     *        means the the 0th list contains the set of columns that are present in
+     *        the 0th grouping.
+     * @return A bit set converted to decimal indicating which columns are present in
+     *         the grouping.
+     *
+     */
+    @ScalarFunction(deterministic = false)
+    @SqlType(StandardTypes.INTEGER)
+    public static long grouping(
+            @SqlType(BIGINT) long groupId,
+            @SqlType("ListLiteral") List groupingOrdinals,
+            @SqlType("ListLiteral") List groupingSetOrdinals)
+    {
+        BitSet groupingBinary = new BitSet(groupingOrdinals.size());
+        groupingBinary.set(0, groupingOrdinals.size());
+        List<Integer> groupingSet = (List<Integer>) groupingSetOrdinals.get((int) groupId);
+        for (Integer groupingColumn : groupingSet) {
+            if (groupingOrdinals.contains(groupingColumn)) {
+                groupingBinary.clear(groupingOrdinals.indexOf(groupingColumn));
+            }
+        }
+
+        int grouping = 0;
+        // Rightmost argument to grouping() is represented by the least significant bit
+        // so we start the conversion from binary to decimal from the left.
+        for (int i = groupingOrdinals.size() - 1, j = 0; i >= 0; i--, j++) {
+            if (groupingBinary.get(i)) {
+                grouping += Math.pow(2, j);
```

Comment:
I think you might have been looking at an outdated version of the PR, I added a commit addressing in person comments where I made this change.